app/controllers/review_mapping_controller.rb

```diff
@@ -473,105 +520,141 @@ def assign_reviewers_for_team(assignment_id, review_strategy, participants_hash)
-            if participants_hash[participant.id] == review_strategy.reviews_per_student
-              participants.delete_at(rand_num)
-              num_participants -= 1
-            end
-          end
-        end
+        even_out_reviews_among_teams(num_participants, maximum_reviews_per_student)
       else
-        # REVIEW: num for last team can be different from other teams.
         # prohibit one student to review his/her own artifact and selected_participants cannot include duplicate num
         participants.each do |participant|
           # avoid last team receives too many peer reviews
-          if !TeamsUser.exists?(team_id: team.id, user_id: participant.user_id) && (selected_participants.size < review_strategy.reviews_per_team)
-            selected_participants << participant.id
-            participants_hash[participant.id] += 1
+          ## why this selected_participants condition since it's empty
+          if !TeamsUser.exists?(team_id: team.id, user_id: participant.user_id) && selected_participants.size < maximum_reviews_per_team
+            modify_selected_participants_to_review(participant.id, selected_participants, participants_hash)
           end
         end
       end
-
       begin
         selected_participants.each { |index| ReviewResponseMap.where(reviewee_id: team.id, reviewer_id: index, reviewed_object_id: assignment_id).first_or_create }
       rescue StandardError
         flash[:error] = 'Automatic assignment of reviewer failed.'
       end
+      iterator += 1
+    end
+  end
+
+  def modify_selected_participants_to_review(participant_id, selected_participants, participants_hash)
+    # selected_participants cannot include duplicate num
+    selected_participants << participant_id
+    participants_hash[participant_id] += 1
+  end
+
+  def get_random_number(iterator, participants_hash, num_participants)
+    if iterator.zero?
+      rand_num = rand(0..num_participants - 1)
+    else
+      rand_num = condition_for_else(participants_hash)
+    end
+    return rand_num
+  end
+
+
+  def number_of_participants_in_team(assignment_id, team)
+    num_participants_this_team = TeamsUser.where(team_id: team.id).size
+    # If there are some submitters or reviewers in this team, they are not treated as normal participants.
+    # They should be removed from 'num_participants_this_team'
+    TeamsUser.where(team_id: team.id).each do |team_user|
+      temp_participant = Participant.where(user_id: team_user.user_id, parent_id: assignment_id).first
+      num_participants_this_team -= 1 if temp_participant.can_review == false or temp_participant.can_submit == false
     end
+    return num_participants_this_team
+  end
+
+  def condition_for_else(participants_hash)
+    participants_with_min_assigned_reviews = participants_with_min_assigned_reviews(participants_hash)
+    # if participants_with_min_assigned_reviews is blank
+    if_condition_1 = participants_with_min_assigned_reviews.empty?
```

Comment:
Please rename variables `if_condition_1` and `if_condition_2` to a more informative ones.